drivers/spi/spi_mchp_mec5_qspi.c

```diff
@@ -0,0 +1,729 @@
+		return -ENOTSUP;
+	}
+
+	if (SPI_WORD_SIZE_GET(config->operation) != 8) {
+		LOG_ERR("Word size != 8 not supported");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static enum mec_qspi_io lines_to_io(uint32_t opw)
+{
+	enum mec_qspi_io iom = MEC_QSPI_IO_FULL_DUPLEX;
+
+#if defined(CONFIG_SPI_EXTENDED_MODES)
+	switch (opw & SPI_LINES_MASK) {
+	case SPI_LINES_DUAL:
+		iom = MEC_QSPI_IO_DUAL;
+		break;
+	case SPI_LINES_QUAD:
+		iom = MEC_QSPI_IO_QUAD;
+		break;
+	default:
+		break;
+	}
+#endif
+	return iom;
+}
+
+static int mec5_qspi_configure(const struct device *dev, const struct spi_config *config)
+{
+	const struct mec5_qspi_config *devcfg = dev->config;
+	struct mec_qspi_regs *regs = devcfg->regs;
+	struct mec5_qspi_data *data = dev->data;
+	uint8_t sgm = 0;
+	int ret = 0;
+
+	if (!config) {
+		return -EINVAL;
+	}
+
+	/* chip select */
+	if (config->slave >= MEC_QSPI_CS_MAX) {
+		LOG_ERR("Invalid chip select [0,1]");
+		return -EINVAL;
+	}
+
+	data->cs = (uint8_t)(config->slave & 0xffu);
+	mec_hal_qspi_cs_select(regs, data->cs);
+
+	data->iom = MEC_QSPI_IO_FULL_DUPLEX;
+
+	if (config->frequency != data->freq) {
+		ret = mec_hal_qspi_set_freq(regs, config->frequency);
+		if (ret != MEC_RET_OK) {
+			return -EINVAL;
+		}
+		data->freq = config->frequency;
+		mec_hal_qspi_byte_time_ns(regs, &data->byte_time_ns);
+	}
+
+	if (config->operation == data->operation) {
+		return 0;
+	}
+
+	data->operation = config->operation;
+	ret = spi_feature_support(config);
+	if (ret) {
+		return ret;
+	}
+
+	data->iom = lines_to_io(data->operation);
+
+	ret = mec_hal_qspi_io(regs, data->iom);
+	if (ret != MEC_RET_OK) {
+		return -EINVAL;
+	}
+
+	if (data->operation & SPI_MODE_CPHA) {
+		sgm |= BIT(0);
+	}
+	if (data->operation & SPI_MODE_CPOL) {
+		sgm |= BIT(1);
+	}
+	/* requires QSPI frequency to be programmed first */
+	ret = mec_hal_qspi_spi_signal_mode(regs, mec5_qspi_sig_mode[sgm]);
+	if (ret != MEC_RET_OK) {
+		return -EINVAL;
+	}
+
+	data->ctx.config = config;
+
+	return 0;
+}
+
+static int mec5_qspi_fd(const struct device *dev, const struct spi_buf_set *tx_bufs,
+			const struct spi_buf_set *rx_bufs, bool async)
+{
+	const struct mec5_qspi_config *devcfg = dev->config;
+	struct mec5_qspi_data *data = dev->data;
+	struct mec_qspi_regs *regs = devcfg->regs;
+	struct spi_context *ctx = &data->ctx;
+	size_t xlen = 0, txlen = 0, rxlen = 0, total_tx = 0, total_rx = 0;
+	uint32_t qflags = 0, xflags = 0;
+	int ret = 0;
+	bool hold_cs = (data->operation & SPI_HOLD_ON_CS) ? true : false;
+
+	xflags = MEC_QSPI_ULDMA_FLAG_TX_OVR_VAL_SET(xflags, devcfg->ovrc);
+	xflags |= MEC5_QSPI_ULDMA_FLAG_START | MEC5_QSPI_ULDMA_FLAG_IEN;
+
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, 1u);
+
+	total_tx = spi_context_total_tx_len(ctx);
+	total_rx = spi_context_total_rx_len(ctx);
+	data->total_tx_size = total_tx;
+	data->total_rx_size = total_rx;
+
+	spi_context_cs_control(ctx, true);
+
+	while (spi_context_rx_on(ctx) || spi_context_tx_on(ctx)) {
```

Comment:
If properly interrupt based, this while loop would not be needed. Here it's creating context switches back to caller's thread to fill in the controller with data, but this should be done in isr mode. So you'll need to revise the logic here.

Basically the common logic is:
- you configure
- you fire the interrupt handling
- you wait (on spi_context_wait_for_completion()) 
Rest: filling in and reading out the fifos is done while handling the interrupts.

Reply:
@tbursztyka OK. Will re-architect code putting the logic in the ISR.
I guess I need to create two more drivers for this HW block: mspi and flash for dual/quad support.